src/components/DemoBoardModal/DemoBoardModal.tsx

```diff
@@ -0,0 +1,434 @@
+    const [blackRanking, setBlackRanking] = React.useState<number>(initialSettings.black_ranking);
+    const [whiteRanking, setWhiteRanking] = React.useState<number>(initialSettings.white_ranking);
+    const [komiOption, setKomiOption] = React.useState<rest_api.KomiOption>(
+        initialSettings.komi_auto,
+    );
+    const [komi, setKomi] = React.useState<number | null>(initialSettings.komi ?? null);
+    const [isCustomBoardSize, setIsCustomBoardSize] = React.useState<boolean>(
+        !isStandardBoardSize(`${initialSettings.width}x${initialSettings.height}`),
+    );
+
+    function updateRules(rules: string): void {
+        if (isRuleSet(rules)) {
+            setRuleSet(rules);
+        }
+    }
+    function updateKomiOption(newKomiOption: string): void {
+        if (!isKomiOption(newKomiOption)) {
+            return;
+        }
+
+        const changedToCustom = newKomiOption === "custom" && komiOption !== "custom";
+        if (changedToCustom) {
+            setKomi(getDefaultKomi(ruleSet, false));
+        }
+        setKomiOption(newKomiOption);
+    }
+    function updateBoardSize(str: string) {
+        if (str === "custom") {
+            setIsCustomBoardSize(true);
+            return;
+        }
+
+        setIsCustomBoardSize(false);
+        const sizes = str.split("x");
+        setWidth(parseInt(sizes[0]));
+        setHeight(parseInt(sizes[1]));
+    }
+
+    const user = data.get("user");
+
+    function getValidatedSettings(): DemoSettings | null {
+        if (width === null || width < 1 || width > 25) {
+            document.getElementById("demo-board-modal-width")?.focus();
+            void alert.fire(_("Invalid board size, please correct and try again"));
+            return null;
+        }
+        if (height === null || height < 1 || height > 25) {
+            document.getElementById("demo-board-modal-height")?.focus();
+            void alert.fire(_("Invalid board size, please correct and try again"));
+            return null;
+        }
+        if (komiOption === "custom" && komi === null) {
+            document.getElementById("demo-board-modal-komi")?.focus();
+            void alert.fire(_("Please enter a number for komi."));
+            return null;
+        }
+
+        return {
+            name: name,
+            rules: ruleSet,
+            width: width,
+            height: height,
+            black_name: blackName,
+            black_ranking: blackRanking,
+            white_name: whiteName,
+            white_ranking: whiteRanking,
+            private: isPrivate,
+            komi_auto: komiOption,
+            ...(komi !== null && { komi: komi }),
+        };
+    }
+
+    async function createDemoBoard(): Promise<void> {
+        const validatedSettings = getValidatedSettings();
+        if (validatedSettings === null) {
+            return;
+        }
+
+        data.set("demo.settings", validatedSettings);
+
+        const black_pro = validatedSettings.black_ranking > 1000 ? 1 : 0;
```

Comment:
I'm not sure why this and the following steps are necessary, but I basically took this from the existing code.